src/ncp/spinel_encoder.cpp

```diff
@@ -0,0 +1,277 @@
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ *    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ *    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+ *    DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ *    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ *    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a spinel encoder.
+ */
+
+#include <openthread/config.h>
+
+#include "spinel_encoder.hpp"
+
+#include "common/code_utils.hpp"
+
+namespace ot {
+namespace Ncp {
+
+otError SpinelEncoder::BeginFrame(NcpFrameBuffer::Priority aPriority)
+{
+    mNumOpenStructs = 0;
+    return mNcpBuffer.InFrameBegin(aPriority);
+}
+
+otError SpinelEncoder::BeginFrame(uint8_t aHeader, unsigned int aCommand)
+{
+    otError error = OT_ERROR_NONE;
+
+    // Non-zero TID indicates this is a response to a spinel command.
+
+    if (SPINEL_HEADER_GET_TID(aHeader) != 0)
+    {
+        SuccessOrExit(error = BeginFrame(NcpFrameBuffer::kPriorityHigh));
+    }
+    else
+    {
+        SuccessOrExit(error = BeginFrame(NcpFrameBuffer::kPriorityLow));
+    }
+
+    SuccessOrExit(error = WriteUint8(aHeader));
+    SuccessOrExit(error = WriteUintPacked(aCommand));
+
+exit:
+    return error;
+}
+
+otError SpinelEncoder::BeginFrame(uint8_t aHeader, unsigned int aCommand, spinel_prop_key_t aKey)
+{
+    otError error = OT_ERROR_NONE;
+
+    SuccessOrExit(error = BeginFrame(aHeader, aCommand));
+
+    // The write position is saved before writing the property key,
+    // so that if fetching the property fails and we need to
+    // reply with a `LAST_STATUS` error we can get back to
+    // this saved write position and update the property key.
+    // (Also see `OverwriteWithLastStatusError()`).
+
+    SuccessOrExit(error = SavePosition());
+    SuccessOrExit(error = WriteUintPacked(aKey));
+
+exit:
+    return error;
+}
+
+otError SpinelEncoder::OverwriteWithLastStatusError(spinel_status_t aStatus)
+{
+    otError error = OT_ERROR_NONE;
+
+    SuccessOrExit(error = ResetToSaved());
+    SuccessOrExit(error = WriteUintPacked(SPINEL_PROP_LAST_STATUS));
+    SuccessOrExit(error = WriteUintPacked(aStatus));
```

Comment:
Here, in this instance, you have accomplished what I was describing above.  You have encoded the entire last status message in a method that can be shared by the host.  Conceivably, every NCP message could be encoded (and decoded) here in this file side by side so as to create the best opportunity to identify mistakes.